runtime/include/mem/jemalloc/chpl-mem-impl.h

```diff
@@ -35,57 +35,50 @@
 #define CHPL_JE_TOK_PASTE(a, b) CHPL_JE_TOK_PASTE_INTERNAL(a,b)
 #define CHPL_JE_(NAME) CHPL_JE_TOK_PASTE(CHPL_JEMALLOC_PREFIX, NAME)
 
-#define CHPL_JE_CALLOC CHPL_JE_(calloc)
-#define CHPL_JE_MALLOC CHPL_JE_(malloc)
-#define CHPL_JE_POSIX_MEMALIGN CHPL_JE_(posix_memalign)
-#define CHPL_JE_REALLOC CHPL_JE_(realloc)
-#define CHPL_JE_FREE CHPL_JE_(free)
+#define CHPL_JE_MALLOCX CHPL_JE_(mallocx)
+#define CHPL_JE_RALLOCX CHPL_JE_(rallocx)
+#define CHPL_JE_DALLOCX CHPL_JE_(dallocx)
 #define CHPL_JE_NALLOCX CHPL_JE_(nallocx)
 #define CHPL_JE_MALLCTL CHPL_JE_(mallctl)
 
 
 #define MALLOCX_NO_FLAGS 0
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-extern unsigned get_num_arenas(void);
-extern unsigned set_arena(unsigned);
-#ifdef __cplusplus
+// Determine which arena to use. For large allocations (32 MiB) use a dedicated
+// arena to reduce fragmentation
+extern unsigned CHPL_JE_LG_ARENA;
+static inline int CHPL_JE_MALLOCX_ARENA_FLAG(size_t size) {
+  if (size >= ((size_t) 32 << 20)) {
+    return MALLOCX_ARENA(CHPL_JE_LG_ARENA);
+  }
+  return MALLOCX_NO_FLAGS;
 }
-#endif
+
 
 static inline void* chpl_calloc(size_t n, size_t size) {
-  return CHPL_JE_CALLOC(n,size);
+  size_t nSize = n*size;
+  if (nSize == 0) { nSize = 1; }
```

Comment:
Why not just return NULL, instead of going to the trouble of allocating an unwanted byte?

Reply:
Our memory API interprets a `NULL` return as OOM, so we return a valid pointer. See also https://github.com/chapel-lang/chapel/issues/5273 for some additional detail

Reply:
Oh, of course, thanks.